### PR TITLE
macOS Sierra (10.12): Disallow Automatic Window Tabbing

### PR DESCRIPTION
**Pull request by nelsonjchen**
Before:
screen shot 2017-03-08 at 12 41 39 am
After:
screen shot 2017-03-08 at 12 36 03 am

Automatic Window Tabbing in macOS Sierra causes a menu item called "Show
Tab Bar" to appear under the Window menu in the menu bar. Selecting this
option causes a double-title bar effect as there is no option in OpenRCT2
to make new tabbed "documents". This double-title bar can be untoggled
by selecting "Hide Tab Bar".

screen shot 2017-03-08 at 12 41 46 am

I don't think OpenRCT2 is going to become a tabbed document-oriented
themepark simulator anytime soon. Thanks, but no thanks Sierra.

References:
* https://codereview.chromium.org/2325313002/
* https://hg.mozilla.org/releases/mozilla-aurora/rev/385b9ea08ff7

### CODE REVIEW

**Comment by LRFLEW** [comment]
I looked over this, and overall, I approve.

I tried quickly upgrading the version of SDL (since we're a few versions behind), and it doesn't fix the issue, so this won't be made redundant when I finally update it (I want to do it at the same time as adding Boost). I pulled this version, and it fixed the issue, so it does work.

The concern is how this affects older versions of OS X. It looks like it should work OK on older versions (I can't test it myself since I don't have it installed anywhere), and Xcode doesn't complain about the compatibility flag, so that should be fine. However, the way it's written, it forces the compiler to use the 10.12 SDK (basically forcing it to be compiled on 10.12). It seems reasonable enough to make that requirement if it's needed, but it's a simple change to allow for 10.11 SDK (and lower) to be able to build it (I tested it myself). Since there may still be users who compile and run it on an older version of the OS, it would be nice to not artificially restrict it. I made the code-review comment about what change would be needed. With that change, this gets my :+1:.

**Comment by nelsonjchen** [comment]
Thanks for taking a look. 

It's odd that newer SDL doesn't fix this. I guess it might be worth fixing it upstream. I don't think there's much, if any, SDL applications that can partake in getting anything useful out of this functionality. Will have to figure out their mailing list, mercurial, and so on to get that in.

I noticed that the browser vendor references I was using did also look into the SDK issue. Some of the soon-after followup patches or commits explicitly added guards to allow building with older SDKs. Oddly, the Chromium commit/patch did note that things happened to compile on 10.11 anyway. Eventually Mozilla and Chromium settled on adding a stub when building on older OS versions so this code would build anyway for sure.

Oddly enough, according to Travis, Travis runs the build in Xcode 8 running on 10.11. I don't know what SDK gets automatically selected when doing that though. This might be slipping in under that same thing that the Chromium people saw too. 

Come to think of it, I haven't tested the artifact that comes out of Xcode 8 on 10.11. I wonder if that is still as effective.

**Comment by LRFLEW** [comment]
I did test it. I installed the 10.11 SDK (from [here](https://github.com/phracker/MacOSX-SDKs)) with the latest version of Xcode. With the changes I suggested, it compiled with two additional warnings, but it still removed the tab bar option. I think the way I suggested would be the best for now. @marijnvdwerf If you have thoughts on this, feel free to chime in.

**Comment by nelsonjchen** [comment]
Did you submit the review? I don't see your code review comment.

**Comment by LRFLEW** [comment]
I thought I did.

The short version is I'd replace the one function with this:
```
void macos_disallow_automatic_window_tabbing()
{
	@autoreleasepool {
		if ([NSWindow respondsToSelector:@selector(setAllowsAutomaticWindowTabbing:)]) {
			[NSWindow setAllowsAutomaticWindowTabbing:NO];
		}
	}
}
```

**Comment by LRFLEW** [comment]
Ah, finally figured out this request thing. You couldn't see it because I hadn't "submitted" it or something 😝

File: src/openrct2/platform/macos.m [hunk 1]
```diff
@@ -24,6 +24,15 @@
 #include "../localisation/language.h"
 #include "../config/Config.h"
 
+void macos_disallow_automatic_window_tabbing()
+{
+	@autoreleasepool {
+		if ([NSWindow respondsToSelector:@selector(allowsAutomaticWindowTabbing)]) {
+			NSWindow.allowsAutomaticWindowTabbing = NO;
```
Comment (nelsonjchen) [inline src/openrct2/platform/macos.m]: OK

**Comment by nelsonjchen** [comment]
[Ah yeah, it's a new thing introduced last September.][1] This PR is small but if it were a big one, there may be many more comments across many files and it cuts down on having an email for each section of line and instead just bundles it all into one notification. ❤️ it!

Anyway, I've updated the branch with the requested changes. But, I'm still uncertain as to the reason for the change. I'm very much not great at Obj-C at all.

[1]:https://github.com/blog/2256-a-whole-new-github-universe-announcing-new-tools-forums-and-features

**Comment by janisozaur** [comment]
any reason this is not in yet?

**Comment by marijnvdwerf** [comment]
Cmake output:

```
[352/353] Building C object CMakeFiles/openrct2.dir/src/openrct2/platform/macos.m.o
/Users/Marijn/Projects/OpenRCT2/Main/src/openrct2/platform/macos.m:31:14: warning: class method '+setAllowsAutomaticWindowTabbing:' not found (return type defaults to 'id') [-Wobjc-method-access]
                        [NSWindow setAllowsAutomaticWindowTabbing:NO];
                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```

**Comment by nelsonjchen** [comment]
I think that's probably why the browser vendors went with conditionally producing a stub to suppress that warning in their codebases.

**Comment by marijnvdwerf** [comment]
Any ideas on solving this would be appreciated :). Not sure what minimum SDK we require though.

**Comment by LRFLEW** [comment]
This error should ONLY be appearing if you aren't using the 10.12 SDK. If you are using the latest version of Xcode, you should be using the latest SDK, and won't see this warning. For the record, the code change I suggested for this PR changed an error into this warning.

**Pull request merged**